### ISSUE TITLE
Add support for translucent censor bars

### ISSUE THREAD

**Issue by Tuckyyyy** [deeppomf/DeepCreamPy#119]
In some cases the censor bar is only about 75% opaque and the original image is still visible through it, the censor bar just needs to be returned to the original brightness, is that something this could be trained to do? 

SFW example of the type of censoring I'm referring to: 

![image](https://user-images.githubusercontent.com/49401954/55727727-39db6b80-59d8-11e9-9efd-f16515338fdf.png)

**Comment by CharonM72** [comment]
They're just as commonly white as black, so it should go both ways if this is implementable.

**Comment by Samuel-Clarke123** [comment]
Couldn't this just be done by brightening the region by a 4x factor (assuming 75% or 3/4 opacity)?
You'd lose image detail and effectively rasterise it, but still. This seems more like an image manipulation than an image reconstruction problem.

Otherwise, since it can handle mosaics, it could probably handle this, but that's mostly just speculation.

**Comment by deeppomf** [comment]
I'll put this on my to-do list.
@Samuel-Clarke123 Finding the right factor is hard.

**Comment by ccppoo** [comment]
Someone I know that has hobby translating hentai asked about this feature too. It will be great for 
 translucent speech bubble

**Comment by PureStarch** [comment]
It's a fairly easy problem even analytically and I don't think deep learning is the correct tool for the full processing here.

Translucent bars are most of time made using a simple fill of a black color with partial opacity, e.g. 85% in the image above. This is nothing but a linear interpolation operation where the original color is modified by the usual `r = a * (1 - p) + b * p`.

So our variables in a censored picture consist of the fillcolor and the opacity. The fillcolor is easy to fixate in the reverse equation by finding a before/after pair by

- Comparing a pixel just outside the censor bar and just pixel inside the censor bar which you think were the exact same original color.
- Finding a part which was likely full white inside the censor bar before it was censored.

Now you could just do the following in some HLSL'ish pseudocode

```
// Assume censoredcolor is the result of lerp(originalcolor, fillcolor, opacity)
// and that we know censoredcolor and originalcolor values for any pixel

float opacity = 0.85;
float3 x = originalcolor * (opacity - 1.0) + censoredcolor;
float3 divisorcolor = float3(255, 255, 255) * (1.0 - opacity);
float3 sample = pixels[i, j];
sample = (sample - x) / divisorcolor;
pixels[i, j] = sample;
```

This operation effectively reverses a translucent fill of an arbitrary color. The core problem becomes automatically cleaning up the edges of the censor bar as they're often softened and result in selection artifacts. Censoring also reduces the dynamic range of the image and often results in banding and more visible jpeg artifacts when reversed.

The opacity value must also be eyeballed until you find the one that was used for censoring the image. This could probably be automated too using some sort of statistical analysis going from 1% to 100% opacity in steps since wrong values seem to warp the colors strongly, but I haven't looked into this too much.

However since it's an analytical method the results are pretty much perfect unless there's like a 95% opacity fill and the data just isn't there anymore.

I've written a plugin for Photoshop that does exactly this but it's hindered by Photoshop scripting API being really bad with 32bit mode so it's suffering from 8bit representation inaccuracies, so it's not quite ready for the world. Either way it proves the point, for the image above getting the original color (white) and it's censored color (40,40,40) and using opacity of 0.85 gives the following result:

https://i.imgur.com/eW2ztC8.png

What deep learning could help in this process is probably:
- Automatic search of the censor bars (unreliable)
- Manually find censor bars near user annotated area e.g. click (quite reliable)
- Cleaning of the artifacts such as selection edges and banding from reversing the interpolation

Anyway I don't know how comprehensive uncensoring tool you're planning on making here but I'm all up for it having this kind of de-translucer. Just make it analytical and not imagine the stuff inside if the source data is still there.

PS. Just do Patreon

**Comment by ccppoo** [comment]
@PureStarch 
Thanks for your comment and your research. I also thought about how to detect Translucent bars by statistics not deep learning, and it's really tricky and needs understanding about colors used in art.

What I'm thinking is if we set a premise that color is used naturally (opposite kind of this https://www.deviantart.com/neoartcore/art/KDA-Neon-771829896, my theory might be wrong ) and there's almost no vertical or horizontal lines, finally translucent speech bubble is rectangular.

Means of pixel RGB of constant certain area will show normal distribution. Then if the "Translucent bar detector" meets translucent bar, normal distribution of constant area of RGB will show two means. One is where bright zone with out bars that will have relatively high means of RGB (may be about 150~200). Other one is where the bar covers and will have low means of RGB(20~30% of high ones). If we visualize it, it will show two camel humps(means) slowly as the detector moves maybe 10 pixels at a time, and slowly disappear when detector moves away from the bar.

After finishing the scan, variance of RGB for each constant area will show difference by size of detecting area. (latter part omitted)

If this theory works, selecting a optimized size(exactly ratio) for detecting translucent bar will be trained by deep-learning I guess? and maybe getting data of what opacity used in bars could be done too.

Any opinions welcomed. Will love it if it's about what's wrong about what I thought.

**Comment by PureStarch** [comment]
@ccppoo I also gave a try to finding censor bars using analytical methods in the past, but ended up giving it up after around 1.5k lines of concise, hard to read optimized C++/OpenCV was still giving a lot of false positives and the processing time for making further features was getting out of hand on CPU-side.

These kind of object recognition problems are exactly the kind of problems neural networks excel at. The amount of features that define a censor bar (or a censored area in general) are too vague for solving the problem using analytical/"logic" methods without excessive amount of code / processing time. The amount of non-rectangular bars, partially covered bars, not-quite-rectangle bars, white bars and all the other variants are too numerous to cover using that kind of logical bruteforcing.

However the core issue that really undermines analytical approaches for finding censor bars is that what makes up a censor bar is in the end not the rectangle, but what is around it.

Solving this issue of:

1. Finding the censored parts of an image automatically (object recognition)
2. Creating a large dataset out of censored areas in images
3. Training a network to find uncensored parts with this censored dataset (likely need to modify the training set to ignore the black bar by e.g. content-aware filling the black bar out)
4. Creating a large dataset of uncensored areas in images
5. Training another network on the uncensored dataset using automatically placed black bars

...is probably the way DCP training set is being/should be constructed on the long run. This is hard to automatize 100% and probably needs some manual supervision in generating the datasets.

It's possible DCP is built by the author just manually browsing uncensored images and clipping sections for the dataset, going straight to the point 4, but obviously this approach wouldn't scale forever.

**Comment by ccppoo** [comment]
@PureStarch 
Yup, I almost spent all day to change into codes I knew why this not simple as I thought.
In this case, looking into digital art guide book might give us a clue for setting premise which will make more specific for modeling(reducing numerous kinds of bars). 

If you have concrete models of what you are thinking of, can you share it?

**Comment by deeppomf** [comment]
@PureStarch @ccppoo 
Tried analytic approaches for censor detection. They don't work. Like you've said, there's too much variation in censors.
Looking into pure ML ways for censor detection like semantic segmentation.

**Comment by Zero3K** [comment]
Is there any news regarding this suggestion?
